tests/cases/conformance/types/unknown/unknownControlFlow.ts

```diff
@@ -269,3 +269,35 @@ function foo<T>(x: T | null) {
         y;
     }
 }
+
+// We allow an unconstrained object of a generic type `T` to be indexed by a key of type `keyof T`
+// without a check that the object is non-undefined and non-null. This is safe because `keyof T`
+// is `never` (meaning no possible keys) for any `T` that includes `undefined` or `null`.

```

Comment:
Isn't this sort of an unusual thing to "defer" the error reporting to the call site? I'm not totally sure what I think about this yet but my first thought was that it's a special case in the algorithm that might be hard to explain to people.

Reply:
I guess this build on top of the fact that `T[never]` is always OK and `never` is always assignable to everything so perhaps this fits the existing rules just OK.

Reply:
I added the test and the comment to capture the behavior for future reference. It's nothing new actually. We have the [same behavior in 4.7](https://www.typescriptlang.org/play?ts=4.7.4#code/GYVwdgxgLglg9mABMYBGAPAFQHwAooBcimANIgNZHkCmAnnMMQJSIDeAsAFCI+JQDa5ALoBuLgF8uXUJFgJkwAExY8hRADkE6kABsdAQwBGO6irKUKdBszZdefQaIlTOM6PCQoAzCvxFSFFRWjJpg2npGJiosHNy8AsIiPAD0yYgAogBOmXCZzpzS4O7yKAAsvmqh4QbGpjjmQfQhWro1UTgxdvGOYpySBa5ouGCtZADkwHBwY0y9KIrDo4gTUzNzwF6LeuOT07PSwKVbOjur+wNAA) with the old `NonNullable<T>` based on a conditional type. This PR ensures we carry it forward in 4.8 with `NonNullable<T>` based on `T & {}`.

Reply:
Sure - but I'm not thinking here from the PoV of `NonNullable` at all. My confusion/concern was solely about unconstrained generics and your changes from https://github.com/microsoft/TypeScript/pull/49119, I've assumed that given those changes this indexed access would error nowadays. Since the base constraint of unconstrained `T` is `unknown` it feels off that we can use indexed access on it without refining the type somehow first

Reply:
It may seem surprising at first, but it is actually quite meaningful because we can permit `t[k]` _without_ requiring local proof that `t` is non-undefined and non-null. Such proof would add unnecessary overhead for instantiations where the type argument for `T` is already known not to be undefined or null.